File: calculator.ts
/**
 * CarbonConstruct API Client
 * TypeScript client for FastAPI backend
 */

const API_BASE_URL = import.meta.env.VITE_API_URL || 'http://localhost:8000/api/v1';

export interface FuelCalculationRequest {
  project_id: string;
  fuel_type: string;
  quantity: number;
  unit: string;
  state: string;
  is_stationary: boolean;
}

export interface MaterialCalculationRequest {
  project_id: string;
  material_type: string;
  quantity: number;
  unit: string;
  data_quality?: 'default' | 'avg' | 'min' | 'max';
}

export interface WasteCalculationRequest {
  project_id: string;
  waste_type: string;
  quantity: number;
  unit?: string;
}

export interface FuelCalculationResponse {
  co2e_kg: number;
  breakdown: {
    co2_kg: number;
    ch4_kg: number;
    n2o_kg: number;
  };
  energy_gj: number;
  factor_source: string;
  uncertainty_pct: number;
  compliance: string;
}

export interface MaterialCalculationResponse {
  gross_co2e_kg: number;
  carbon_storage_kg: number;
  net_co2e_kg: number;
  factor_source: string;
  uncertainty_pct: number;
  data_quality: string;
  compliance: string;
}

export interface WasteCalculationResponse {
  co2e_kg: number;
  factor_source: string;
  factor_t_co2e_per_t: number;
  uncertainty_pct: number;
  compliance: string;
}

export interface ProjectSummary {
  project_id: string;
  total_co2e_kg: number;
  total_co2e_tonnes: number;
  breakdown: {
    [key: string]: {
      count: number;
      co2e_kg: number;
      uncertainty_pct?: number;
    };
  };
  recent_calculations: Array<{
    timestamp: string;
    activity: string;
    description: string;
    co2e_kg: number;
  }>;
  timestamp: string;
  compliance: string;
}

export interface APIResponse<T> {
  status: 'success' | 'error';
  data?: T;
  message?: string;
  errors?: string[];
}

export interface MaterialInfo {
  material_type: string;
  material_category: string;
  unit: string;
  a1a3_default_per_unit: number;
  data_quality: string;
  carbon_storage_per_unit?: number;
}

export interface FuelInfo {
  fuel_type: string;
  category: string;
  region: string;
  unit: string;
  total_co2e: number;
  nger_method: string;
}

class CarbonCalculatorAPI {
  private baseURL: string;

  constructor(baseURL: string = API_BASE_URL) {
    this.baseURL = baseURL;
  }

  private async request<T>(
    endpoint: string,
    options: RequestInit = {}
  ): Promise<APIResponse<T>> {
    const url = `${this.baseURL}${endpoint}`;
    
    const response = await fetch(url, {
      ...options,
      headers: {
        'Content-Type': 'application/json',
        ...options.headers,
      },
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'API request failed');
    }

    return response.json();
  }

  // ========== CALCULATOR ENDPOINTS ==========

  async calculateFuel(data: FuelCalculationRequest): Promise<FuelCalculationResponse> {
    const response = await this.request<FuelCalculationResponse>(
      '/calculate/fuel',
      {
        method: 'POST',
        body: JSON.stringify(data),
      }
    );
    return response.data!;
  }

  async calculateMaterial(data: MaterialCalculationRequest): Promise<MaterialCalculationResponse> {
    const response = await this.request<MaterialCalculationResponse>(
      '/calculate/material',
      {
        method: 'POST',
        body: JSON.stringify(data),
      }
    );
    return response.data!;
  }

  async calculateWaste(data: WasteCalculationRequest): Promise<WasteCalculationResponse> {
    const response = await this.request<WasteCalculationResponse>(
      '/calculate/waste',
      {
        method: 'POST',
        body: JSON.stringify(data),
      }
    );
    return response.data!;
  }

  async listMaterials(category?: string, search?: string): Promise<MaterialInfo[]> {
    const params = new URLSearchParams();
    if (category) params.append('category', category);
    if (search) params.append('search', search);
    
    const response = await this.request<MaterialInfo[]>(
      `/calculate/materials?${params.toString()}`
    );
    return response.data!;
  }

  async listFuels(state?: string, category?: string): Promise<FuelInfo[]> {
    const params = new URLSearchParams();
    if (state) params.append('state', state);
    if (category) params.append('category', category);
    
    const response = await this.request<FuelInfo[]>(
      `/calculate/fuels?${params.toString()}`
    );
    return response.data!;
  }

  async listMaterialCategories(): Promise<string[]> {
    const response = await this.request<string[]>('/calculate/categories');
    return response.data!;
  }

  // ========== PROJECT ENDPOINTS ==========

  async createProject(data: {
    project_id: string;
    project_name: string;
    postcode?: number;
    state?: string;
    ncc_volume?: string;
  }): Promise<any> {
    const response = await this.request('/projects', {
      method: 'POST',
      body: JSON.stringify(data),
    });
    return response.data;
  }

  async getProject(projectId: string): Promise<any> {
    const response = await this.request(`/projects/${projectId}`);
    return response.data;
  }

  async getProjectSummary(projectId: string): Promise<ProjectSummary> {
    const response = await this.request<ProjectSummary>(
      `/projects/${projectId}/summary`
    );
    return response.data!;
  }

  async getAuditLog(projectId: string): Promise<any[]> {
    const response = await this.request<any[]>(`/projects/${projectId}/audit`);
    return response.data!;
  }

  async listProjects(state?: string, nccVolume?: string): Promise<any[]> {
    const params = new URLSearchParams();
    if (state) params.append('state', state);
    if (nccVolume) params.append('ncc_volume', nccVolume);
    
    const response = await this.request<any[]>(`/projects?${params.toString()}`);
    return response.data!;
  }

  // ========== REPORTS ENDPOINTS ==========

  async exportNGERReport(projectId: string): Promise<any> {
    const response = await this.request(`/reports/${projectId}/nger-json`);
    return response.data;
  }

  async exportNCCSummary(projectId: string): Promise<any> {
    const response = await this.request(`/reports/${projectId}/ncc-summary`);
    return response.data;
  }

  async getMethodologyStatement(projectId: string): Promise<any> {
    const response = await this.request(`/reports/${projectId}/methodology`);
    return response.data;
  }
}

// Export singleton instance
export const calculatorAPI = new CarbonCalculatorAPI();

// Export class for custom instances
export default CarbonCalculatorAPI;

File: calculator.py
"""
Calculator API endpoints
"""

from fastapi import APIRouter, HTTPException
from typing import List

from ..models.schemas import (
    FuelCalculationRequest, FuelCalculationResponse,
    MaterialCalculationRequest, MaterialCalculationResponse,
    WasteCalculationRequest, WasteCalculationResponse,
    APIResponse, MaterialInfo, FuelInfo
)
from ...core.calculator_engine import CarbonCalculator
from ...core.database import get_db_connection

router = APIRouter()


@router.post("/fuel", response_model=APIResponse)
async def calculate_fuel(request: FuelCalculationRequest):
    """
    Calculate Scope 1 emissions from fuel combustion
    
    **Critical**: Diesel requires `is_stationary` flag
    - N₂O factors differ by 171% between stationary and transport
    - UI must ask: "Is this fuel used in a registered road vehicle?"
    
    **State-specific factors**: 
    - Electricity and gas factors vary significantly by state
    - Always provide accurate state code
    """
    try:
        calc = CarbonCalculator()
        result = calc.calculate_fuel_emissions(
            project_id=request.project_id,
            fuel_type=request.fuel_type,
            quantity=request.quantity,
            unit=request.unit,
            state=request.state,
            is_stationary=request.is_stationary
        )
        return APIResponse(status="success", data=result)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Calculation error: {str(e)}")


@router.post("/material", response_model=APIResponse)
async def calculate_material(request: MaterialCalculationRequest):
    """
    Calculate A1-A3 embodied carbon from materials
    
    **Data Quality Tiers**:
    - `default`: NGER standard factor (Tier 3)
    - `avg`: Average of available EPDs (Tier 2)
    - `min`: Conservative lower bound (Tier 1)
    - `max`: Conservative upper bound (Tier 1)
    
    **Biogenic Carbon**:
    - Timber includes carbon storage (negative emissions)
    - Net emissions = Gross emissions + Storage (storage is negative)
    """
    try:
        calc = CarbonCalculator()
        result = calc.calculate_material_emissions(
            project_id=request.project_id,
            material_type=request.material_type,
            quantity=request.quantity,
            unit=request.unit,
            data_quality=request.data_quality
        )
        return APIResponse(status="success", data=result)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Calculation error: {str(e)}")


@router.post("/waste", response_model=APIResponse)
async def calculate_waste(request: WasteCalculationRequest):
    """
    Calculate Scope 3 emissions from waste disposal
    
    **Critical**: Timber waste uses DOC method
    - Timber/wood: 1.5 t CO₂e per tonne
    - Concrete: 0.05 t CO₂e per tonne (inert)
    - Steel: 0.1 t CO₂e per tonne
    - Generic organic: 0.2 t CO₂e per tonne
    """
    try:
        calc = CarbonCalculator()
        result = calc.calculate_waste_emissions(
            project_id=request.project_id,
            waste_type=request.waste_type,
            quantity=request.quantity,
            unit=request.unit
        )
        return APIResponse(status="success", data=result)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Calculation error: {str(e)}")


@router.get("/materials", response_model=APIResponse)
async def list_materials(
    category: str = None,
    search: str = None
):
    """
    List available materials with emission factors
    
    Query Parameters:
    - `category`: Filter by material category (e.g., "Concrete", "Steel", "Timber")
    - `search`: Search in material type name
    """
    try:
        db = get_db_connection()
        cursor = db.cursor()
        
        query = "SELECT * FROM nger_materials WHERE 1=1"
        params = []
        
        if category:
            query += " AND material_category = ?"
            params.append(category)
        
        if search:
            query += " AND material_type LIKE ?"
            params.append(f"%{search}%")
        
        query += " ORDER BY material_category, material_type"
        
        materials = cursor.execute(query, params).fetchall()
        db.close()
        
        return APIResponse(
            status="success",
            data=[
                {
                    "material_type": m['material_type'],
                    "material_category": m['material_category'],
                    "unit": m['unit'],
                    "a1a3_default_per_unit": m['a1a3_default_per_unit'],
                    "data_quality": m['data_quality'],
                    "carbon_storage_per_unit": m['carbon_storage_per_unit']
                }
                for m in materials
            ]
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Database error: {str(e)}")


@router.get("/fuels", response_model=APIResponse)
async def list_fuels(
    state: str = None,
    category: str = None
):
    """
    List available fuels with emission factors
    
    Query Parameters:
    - `state`: Filter by state (NSW, VIC, QLD, SA, WA, TAS, NT, ACT)
    - `category`: Filter by category (Stationary, Transport)
    """
    try:
        db = get_db_connection()
        cursor = db.cursor()
        
        query = "SELECT * FROM nger_operational_factors WHERE scope = 1"
        params = []
        
        if state:
            query += " AND region = ?"
            params.append(state.upper())
        
        if category:
            query += " AND category = ?"
            params.append(category)
        
        query += " ORDER BY category, fuel_type, region"
        
        fuels = cursor.execute(query, params).fetchall()
        db.close()
        
        return APIResponse(
            status="success",
            data=[
                {
                    "fuel_type": f['fuel_type'],
                    "category": f['category'],
                    "region": f['region'],
                    "unit": f['unit'],
                    "total_co2e": f['total_co2e'],
                    "nger_method": f['nger_method']
                }
                for f in fuels
            ]
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Database error: {str(e)}")


@router.get("/categories", response_model=APIResponse)
async def list_categories():
    """
    List all material categories for filtering
    """
    try:
        db = get_db_connection()
        cursor = db.cursor()
        
        categories = cursor.execute("""
            SELECT DISTINCT material_category 
            FROM nger_materials 
            ORDER BY material_category
        """).fetchall()
        
        db.close()
        
        return APIResponse(
            status="success",
            data=[c['material_category'] for c in categories]
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Database error: {str(e)}")
